relay/relay.go

```diff
@@ -263,261 +238,173 @@ func (r *Relay) process(parentCtx context.Context) error {
+				latestPos, result.LatestPos, latestGTID, result.LatestGTIDs)
 		}
 
-		ctx, cancel := context.WithTimeout(parentCtx, eventTimeout)
+	}
+	return errors.Annotatef(err, "recover for UUID %s with config %+v", uuid, cfg)
+}
+
+// handleEvents handles binlog events, including:
+//   1. read events from upstream
+//   2. transform events
+//   3. write events into relay log files
+//   4. update metadata if needed
+func (r *Relay) handleEvents(ctx context.Context, reader2 reader.Reader, transformer2 transformer.Transformer, writer2 writer.Writer) error {
+	var (
+		_, lastPos  = r.meta.Pos()
+		_, lastGTID = r.meta.GTID()
+	)
+
+	for {
+		// 1. reader events from upstream server
 		readTimer := time.Now()
-		var e *replication.BinlogEvent
-		if gapReader != nil {
-			e, err = gapReader.GetEvent(ctx)
-		} else {
-			e, err = r.reader.GetEvent(ctx)
-		}
-		cancel()
+		rResult, err := reader2.GetEvent(ctx)
 		binlogReadDurationHistogram.Observe(time.Since(readTimer).Seconds())
 
 		if err != nil {
 			switch errors.Cause(err) {
 			case context.Canceled:
 				return nil
-			case context.DeadlineExceeded:
-				log.Infof("[relay] deadline %s exceeded, no binlog event received", eventTimeout)
-				continue
 			case replication.ErrChecksumMismatch:
 				relayLogDataCorruptionCounter.Inc()
 			case replication.ErrSyncClosed, replication.ErrNeedSyncAgain:
-				// do nothing
+				// do nothing, but the error will be returned
 			default:
 				if utils.IsErrBinlogPurged(err) {
-					if gapReader != nil {
-						pos, err2 := tryFindGapStartPos(err, r.fd.Name())
-						if err2 == nil {
-							log.Errorf("[relay] %s", err.Error())         // log the error
-							gapSyncEndPos2 := *gapSyncEndPos              // save the endPos
-							closeGapSyncer()                              // close the previous gap streamer
-							_, err2 = r.fd.Seek(int64(pos), io.SeekStart) // seek to a valid pos
-							if err2 != nil {
-								return errors.Annotatef(err, "seek %s to pos %d", r.fd.Name(), pos)
-							}
-							err2 = r.fd.Truncate(int64(pos)) // truncate the incomplete event part
-							if err2 != nil {
-								return errors.Annotatef(err, "truncate %s to size %d", r.fd.Name(), pos)
-							}
-							log.Infof("[relay] truncate %s to size %d", r.fd.Name(), pos)
-							gapSyncStartPos = &mysql.Position{
-								Name: lastPos.Name,
-								Pos:  pos,
-							}
-							gapSyncEndPos = &gapSyncEndPos2 // reset the endPos
-							log.Infof("[relay] adjust gap streamer's start pos to %d", pos)
-							continue
-						}
-						return errors.Annotatef(err, "gap streamer")
-					}
-					if tryReSync && r.cfg.EnableGTID && r.cfg.AutoFixGTID {
-						err = r.reSetUpReader(r.syncerCfg)
-						if err != nil {
-							return errors.Annotatef(err, "try auto switch with GTID")
-						}
-						tryReSync = false // do not support repeat try re-sync
-						continue
-					}
+					// TODO: try auto fix GTID, and can support auto switching between upstream server later.
```

Comment:
do we need to add some log here

Reply:
added some log in https://github.com/pingcap/dm/pull/171/commits/f9b1663d8a755052f0307248dd1f9a62ead84dc5.